src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceMarginViewMargin.cs

```diff
@@ -0,0 +1,238 @@
+                streamingFindUsagesPresenter,
+                classificationTypeMap,
+                classificationFormatMap,
+                operationExecutor,
+                editorFormatMap,
+                listener,
+                _mainCanvas);
+            _refreshAllGlyphs = true;
+            _disposed = false;
+
+            _tagAggregator.BatchedTagsChanged += OnTagsChanged;
+            _textView.LayoutChanged += OnLayoutChanged;
+            _textView.ZoomLevelChanged += OnZoomLevelChanged;
+            _optionService.OptionChanged += OnRoslynOptionChanged;
+
+            _mainCanvas.Width = HeightAndWidthOfMargin;
+
+            _grid.LayoutTransform = new ScaleTransform(
+                scaleX: _textView.ZoomLevel / 100,
+                scaleY: _textView.ZoomLevel / 100);
+            _grid.LayoutTransform.Freeze();
+            UpdateMarginVisibility();
+        }
+
+        private void OnZoomLevelChanged(object sender, ZoomLevelChangedEventArgs e)
+        {
+            _grid.LayoutTransform = e.ZoomTransform;
+            _refreshAllGlyphs = true;
+        }
+
+        private void OnLayoutChanged(object sender, TextViewLayoutChangedEventArgs e)
+        {
+            _glyphManager.SetSnapshotAndUpdate(
+                _textView.TextSnapshot,
+                e.NewOrReformattedLines,
+                e.VerticalTranslation ? _textView.TextViewLines : e.TranslatedLines);
+
+            IList<ITextViewLine> lines = _refreshAllGlyphs ? _textView.TextViewLines : e.NewOrReformattedLines;
+            foreach (var line in lines)
+            {
+                _glyphManager.RemoveGlyph(line.Extent);
+                RefreshGlyphsOver(line);
+            }
+
+            _refreshAllGlyphs = false;
+        }
+
+        private void OnRoslynOptionChanged(object sender, OptionChangedEventArgs e)
+        {
+            if (e.Option == FeatureOnOffOptions.ShowInheritanceMargin || e.Option == FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin)
+            {
+                UpdateMarginVisibility();
+            }
+        }
+
+        private void UpdateMarginVisibility()
+        {
+            var featureEnabled = _optionService.GetOption(FeatureOnOffOptions.ShowInheritanceMargin, _languageName) != false;
+            var showMargin = !_optionService.GetOption(FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin, _languageName);
+            if (showMargin && featureEnabled)
+            {
+                _mainCanvas.Visibility = Visibility.Visible;
+            }
+            else
+            {
+                _mainCanvas.Visibility = Visibility.Collapsed;
+            }
+        }
+
+        private void OnTagsChanged(object sender, BatchedTagsChangedEventArgs e)
+        {
+            if (_textView.IsClosed)
+            {
+                return;
+            }
+
+            var changedSnapshotSpans = e.Spans.SelectMany(span => span.GetSpans(_textView.TextSnapshot)).ToImmutableArray();
```

Comment:
consider non-allocating version of this.

Reply:
Do you mean using the PooledArray? 
Sure will do